### ISSUE TITLE
Only use and depend on `patchelf` if it is actually needed

### ISSUE THREAD

**Issue by rgommers** [FFY00/meson-python#125]
Sage ran into an issue where `patchelf` is built from source and break with GCC 6.3 on Debian when trying to use `meson-python` with SciPy (see https://trac.sagemath.org/ticket/34081, Cc @mkoeppe). This is a little unfortunate, because `patchelf` is not actually needed. What it's used for here is to add a non-existing folder to the RPATH - which is not useful:

```python
        # fix file
        if platform.system() == 'Linux':
            # add .mesonpy.libs to the RPATH of ELF files
            if self._is_elf(os.fspath(origin)):
                # copy ELF to our working directory to avoid Meson having to regenerate the file
                new_origin = self._libs_build_dir / pathlib.Path(origin).relative_to(self._project._build_dir)
                os.makedirs(new_origin.parent, exist_ok=True)
                shutil.copy2(origin, new_origin)
                origin = new_origin
                # add our in-wheel libs folder to the RPATH
                elf = mesonpy._elf.ELF(origin)
                libdir_path = f'$ORIGIN/{os.path.relpath(f".{self._project.name}.mesonpy.libs", destination.parent)}'
                if libdir_path not in elf.rpath:
                    elf.rpath = [*elf.rpath, libdir_path]
```

Of course an issue is that `meson-python` cannot currently know when setting up the build environment whether the project has standalone compiled libraries that are going to be included in `.mesonpy.libs`. This would be easy to make controllable via a boolean flag in `pyproject.toml` though - and then the vast majority of Python packages that do not have such libraries don't have to deal with the unnecessary `patchelf` dependency.

It seems to me like it's time for a `[meson-python]` table in `pyproject.toml`, with at least two entries: one for "has standalone libraries" and one for "pure" to avoid the fragile heuristics around that.

**Comment by rgommers** [comment]
This turns out to result in build failures for Linux packagers, because `rpmlint` chokes on the unnecessary RPATH addition - see the report for OpenSUSE at https://github.com/FFY00/meson-python/pull/26#issuecomment-1243265270.

This page explains the issue and Debian's policy pretty well: https://wiki.debian.org/RpathIssue.

We still have no use case for doing this vendoring AFAIK, and it causes real-world problems. So we can do one of two things here:
1. Make it optional (my suggestion above),
2. Remove the feature completely, do a release, and worry about this later.
 
Given that it doesn't work anyway on macOS and Windows, the feature isn't really useful for any Python package right now unless it's a Linux-only package (which very few packages are). So I think (2) is the more logical route here.

**Comment by rgommers** [comment]
@bnavigator a question: it looks like `numpy.distutils` sets RPATH as well, to `<prefix>/lib`. I imagine that there's a way in build tooling of Linux distros to strip rpath's to directories that are searched by default by the linker. And that's why you don't see an issue with $ORIGIN/../lib` while `$ORIGIN/../../.scipy.mesonpy.libs` is problematic. Is that correct, and if so what is that tooling?

The Debian page says that preferably the build tool provides an option to completely remove rpath entries - but Meson doesn't provide that, and neither does `distutils`. So I assume that that's not really needed.

**Comment by bnavigator** [comment]
RPM based distributions check for this in the build process:

Fedora: https://fedoraproject.org/wiki/Changes/Broken_RPATH_will_fail_rpmbuild
openSUSE: https://en.opensuse.org/openSUSE:Packaging_checks#Beware_of_Rpath

The current package for scipy 1.8.1 using numpy.distutils seems to set it for the gfortran calls, but in the end the libraries don't have the RPATH set and openSUSE's rpmlint does not complain about it. Not sure why.
https://build.opensuse.org/build/devel:languages:python:numeric/openSUSE_Tumbleweed/x86_64/python-scipy:standard/_log

**Comment by eli-schwartz** [comment]
> The Debian page says that preferably the build tool provides an option to completely remove rpath entries - but Meson doesn't provide that, and neither does `distutils`. So I assume that that's not really needed.

Not entirely clear what this is even talking about. Meson adds builddir rpaths where needed to make programs work while uninstalled (cmake does this too), but Meson doesn't have an option to refrain from adding build rpaths. Meson also doesn't provide an option to *refrain* from removing build rpath entries at install time, it always removes them.

Meson doesn't add install rpaths on its own. There's a kwarg for adding one, but it's intended to be used solely by projects which know they have a special-cased need -- the standard example is software that installs private shared libraries to `/usr/lib/foobar/libsomething.so.1`, and libsomething isn't intended to be used externally so it isn't installed to the system path.

How this ties into python modules is a different story. The answer, I have always thought, is that this is auditwheel's job.

**Comment by rgommers** [comment]
> Meson doesn't add install rpaths on its own.

I was seeing rpath additions like:
```
$ readelf -d build-install/lib/python3.10/site-packages/scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so | grep rpath
 0x000000000000000f (RPATH)              Library rpath: [/home/rgommers/anaconda3/envs/scipy-dev/lib]
```
so I checked where they came from. It's indeed not Meson, but the conda compilers activation scripts. E.g:
```bash
$ cat ~/anaconda3/envs/scipy-dev/etc/conda/activate.d/activate-gcc_linux-64.sh | grep rpath
# The compiler adds $PREFIX/lib to rpath, so it's better to add -L and -isystem  as well.
  LDFLAGS_USED="-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,${PREFIX}/lib -Wl,-rpath-link,${PREFIX}/lib -L${PREFIX}/lib"
  LDFLAGS_USED="-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,${CONDA_PREFIX}/lib -Wl,-rpath-link,${CONDA_PREFIX}/lib -L${CONDA_PREFIX}/lib"
```

> How this ties into python modules is a different story. The answer, I have always thought, is that this is auditwheel's job.

`auditwheel` is only meant for producing `manylinux`-compliant wheels. It will rewrite absolute to relative rpaths (this would make sense more generally), but it also vendors dependencies that are not in the manylinux standard (never what you want unless you are producing wheels).

**Comment by rgommers** [comment]
Another, more serious, issue caused by the spurious RUNPATH entry: https://github.com/scipy/scipy/issues/17070. The open PR for this issue should fix the problem.

**Comment by rgommers** [comment]
The serious issues with rpath/runpath were fixed here, and the 0.9.0 release containing the fix will follow very soon.

I will open a follow-up issue for the non-urgent original topic of removing the dependency on `patchelf` when it is not needed.
